### ISSUE TITLE
There's no way to call rebalance hooks of StreamConsmer without consuming message

### ISSUE THREAD

**Issue by ik1ne** [fede1024/rust-rdkafka#738]
I'm trying to use `split_partition_queue` of `StreamConsumer`.

I defined a custom `ConsumerContext` which sends revoke/assign event to channel using overridden pre/post-rebalance methods, so a new split partition queue can be created when a partition is assigned.

However, when reassign occurs these hooks aren't being called.

I suspect this is because the function `poll` is not being called after reassign, because awaiting on the stream of the consumer or calling `unsafe { rdkafka_sys::rd_kafka_poll(consumer.client().native_ptr(), 0) };` makes the hooks properly called.

Obviously unsafe call to raw client is not ideal, and awaiting on the stream makes us to handle that message elsewhere - which makes it more complex to integrate it with `split_partition_queue`.

Is there a way to properly invoke rebalance hooks without consuming a message?

**Comment by ik1ne** [comment]
While the test examples has the usage of `split_partition_queue`, they run under specific assumption - the number of partition and its assignment is known in the compile time, allowing the code to hardcode partition number for split partition queue.
I want something that can deal with unknown number of partition and assignment.

**Comment by john-z-yang** [comment]
Hi, I'm not an expert with kafka and this library, so please take what I say with a grain of salt.
But did you spawn a task to await on the consumer after you split from? Like what the doc suggested:

```rust
// Somewhere after calling stream_consumer.split_partition_queue
tokio::spawn(async move {
    let message = stream_consumer.recv().await; // ← poll is being called here to serve rebalancing
    panic!("main stream consumer queue unexpectedly received message: {:?}", message);
})
```

For an example that deal with unknown number of partition and assignment, maybe you can use mine:
https://github.com/john-z-yang/kafka-map-reduce

**Comment by ik1ne** [comment]
> Hi, I'm not an expert with kafka or this library, so please take what I say with a grain of salt. But did you spawn a task to await on the consumer after you called `split_partition_queue`? Like what the doc suggested:
> 
> ```rust
> // Somewhere after calling stream_consumer.split_partition_queue
> tokio::spawn(async move {
>     let _ = stream_consumer.recv().await; // ← poll is being called here to serve things like rebalancing
>     panic!("main stream consumer queue unexpectedly received message: {:?}", message);
> })
> ```
> 
> You should not be getting any messages from the consumer if you always call `split_partition_queue` during a rebalance call back.
> 
> So start the task like above. When a rebalance assignment happens, what your pre-rebalance callback can do is send a oneshot sender through the channel, along with the topic partition list (TPL). Then, block on the oneshot receiver. The task that receives the TPL and oneshot sender can then call `split_partition_queue`, and setup the tasks that handles consuming from each partition. Once that is done, it can drop the oneshot sender to signal to the call back that the partition queue is ready. At which point your call back will unblock and return.
> 
> For an example that uses `split_partition_queue` and can deal with unknown number of partition and assignment, maybe you can use mine: https://github.com/john-z-yang/kafka-map-reduce
> 
> Which is based off of the kafka source from vector.dev https://github.com/vectordotdev/vector/blob/master/src/sources/kafka.rs

Thanks for the reply, this is what I've had in mind as my first implementation but unfortunately it didn't work as expected.

To call `split_partition_queue`, you need to know what partition the consumer has been assigned to. For this I've called `fn assignment(&self) -> KafkaResult<TopicPartitionList>;`. However, this also returns empty TopicPartitonList if `recv()` hasn't been called. In our use case we don't manually assign partition to consumer, so we don't know what partition will be assigned to consumer.

**Comment by john-z-yang** [comment]
No worries!

> To call split_partition_queue, you need to know what partition the consumer has been assigned to. For this I've called fn assignment(&self) -> KafkaResult<TopicPartitionList>;

Couldn't you also get the TPL using the pre-rebalance callback? When your consumer first start up, spawn the task that runs
```rust
stream_consumer.recv().await;
``` 
Then pre-balance will be called with `Rebalance::Assign(&'a TopicPartitionList)` and you can use that to call `split_partition_queue ` using the technique I described above?

**Comment by ik1ne** [comment]
Sorry, I've misread your first comment.

I forgot to mention that our kafka topic might have some unhandled messages before subscription. This is because some producers may produce some messages while the consumers are down. In this case, if we spawn a task that runs `recv()`, it immediately returns some message(before rebalance hooks are called).

**Comment by john-z-yang** [comment]
Oh interesting, I've never observe that before. I have also have tried running my consumer with a topic that already has messages. It always receives a pre-balance callback before messages starts showing up.

How are you synchronising this? Are you blocking in your pre-balance function to make sure that it only returns _after_ the `split_partition_queue ` is called, and .recv().await are being called on each `StreamPartitionQueue`?

**Comment by ik1ne** [comment]
I realized that I was calling `split_partition_queue` on the **post**-rebalance callback, and there might be some synchronization issues.

Thanks for the help! I'll try to call it on the pre rebalance callback, and then check the synchronization logic and will share the result.

**Comment by john-z-yang** [comment]
No problem, best of luck! Feel free to reference what I did here for the specifics!
https://github.com/john-z-yang/kafka-map-reduce

**Comment by ik1ne** [comment]
I reviewed my code and found out these differences:
1. I was handling assignment rebalance events inside post rebalance hook.
2. I was tokio spawning the split partition queue task but wasn't awaiting their `split_partition_queue` calls.
So I moved the event handling from post to pre hook, and added a channel for waiting split partition queue calls. And I have verified that this has solved the main consumer recv call receiving a message.
@john-z-yang Thank you for your help!
